cirq/ops/display.py

```diff
@@ -0,0 +1,175 @@
+
+import numpy as np
+
+from cirq import protocols, value
+from cirq.ops import raw_types, op_tree
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    from cirq.ops import pauli_string
+
+
+class SamplesDisplay(raw_types.Operation):
+    """A display whose value is computed from measurement results.
+
+    The value of a SamplesDisplay on some qubits is computed in the following
+    steps:
+        1. Perform some unitary operations on the qubits
+        2. Sample bitstrings from the resulting state a number of times
+        3. Apply a function to the sampled bitstrings
+    """
+
+    @property
+    @abc.abstractmethod
+    def key(self) -> Hashable:
+        pass
+
+    @abc.abstractmethod
+    def measurement_basis_change(self) -> op_tree.OP_TREE:
+        """Operations to perform prior to measurement."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def repetitions(self) -> int:
+        """The number of measurement samples to take."""
+        pass
+
+    @abc.abstractmethod
+    def value(self,
+              measurements: np.ndarray) -> Any:
+        """The value of the display."""
+        pass
+
+
+class WaveFunctionDisplay(raw_types.Operation):
+    """A display whose value is computed from the full wavefunction."""
+
+    @property
+    @abc.abstractmethod
+    def key(self) -> Hashable:
+        pass
+
+    @abc.abstractmethod
+    def value(self,
+              state: np.ndarray,
+              qubit_map: Dict[raw_types.QubitId, int]
+              ) -> Any:
+        pass
+
+
+@value.value_equality
+class ApproxPauliStringExpectation(SamplesDisplay):
+    """Approximate expectation value of a Pauli string."""
+
+    def __init__(self,
+                 pauli_string: 'pauli_string.PauliString',
+                 repetitions: int,
+                 key: Hashable=''):
+        self._pauli_string = pauli_string
+        self._repetitions = repetitions
+        self._key = key
+
+    @property
+    def qubits(self) -> Tuple[raw_types.QubitId, ...]:
+        return self._pauli_string.qubits
+
+    def with_qubits(self,
+                    *new_qubits: raw_types.QubitId
+                    ) -> 'ApproxPauliStringExpectation':
+        return ApproxPauliStringExpectation(
+                self._pauli_string.with_qubits(*new_qubits),
+                self._repetitions,
+                self._key
+        )
+
+    @property
+    def key(self) -> Hashable:
+        return self._key
+
+    def measurement_basis_change(self) -> op_tree.OP_TREE:
+        return self._pauli_string.to_z_basis_ops()
+
+    @property
+    def repetitions(self) -> int:
+        return self._repetitions
+
+    def value(self,
+              measurements: np.ndarray) -> float:
+        return np.mean([(-1)**sum(bitstring) for bitstring in measurements])
+
+    def _value_equality_values_(self):
+        return self._pauli_string, self._repetitions, self._key
+
+
+@value.value_equality
+class PauliStringExpectation(WaveFunctionDisplay):
```

Comment:
It feels like these two pauli displays are actually the same display with slightly different options. In particular, it seems like it's a kind of "graceful degradation" display where in the simulator you might want idealized values but in the real hardware you want sampled values.

Maybe we could have a method like this:

```python
def pauli_vector_expectation(
        pauli_vector: Union[PauliVector, ..., ..., ...],
        *,
        samples: int,
        fake_statistical_noise_in_simulator: bool = True):
    ....
```

?

Reply:
During a simulation you could want either the idealized value or an approximate value from samples. On hardware you have no choice; you can only get the latter.

Are you proposing that `pauli_vector_expectation` return one of these two displays based on the setting of `fake_statistical_noise_in_simulator`? But then the `samples` argument doesn't have any meaning if we're asking for the idealized wavefunction display, so it's weird to force the user to enter that. We could make it default to 0 or None to indicate the ideal one, and make a different value indicate the samples one.

Reply:
Hmmmm. How about

```python
def pauli_vector_expectation(
        pauli_vector: Union[PauliVector, ..., ..., ...],
        *,
        samples: Optional[int] = None,
        idealized: bool = False):
```

If the user specifies only `samples`, then the simulator adds fake noise and it works on hardware. If the user specifies only `idealized=True` then the simulator doesn't add noise and it doesn't work on hardware. If the user specifies both then the simulator doesn't add noise and also it works on hardware.

I want to push users towards thinking "samples works everywhere so I want to use that, but sometimes I will additionally specify idealized when testing".

Reply:
I think the main alternative is to default to idealized, and if the user specifies samples then we can run on hardware but also inject noise in the simulator. But from the user's perspective I really want them to think of this thing as one object with different options as opposed to two different objects.

Reply:
What if the user specifies neither?

Reply:
I'm leaning toward defaulting to idealized. There would be one keyword argument, `samples`, and specifying a value for it would cause it to return the SamplesDisplay.

Reply:
I've defined the helper function.